build-scripts/lerna-publish.js

```diff
@@ -0,0 +1,144 @@
+  findVersionedUpdates() {
+    let chain = Promise.resolve();
+
+    if (this.options.bump === "from-git") {
+      chain = chain.then(() => this.detectFromGit());
+    } else if (this.options.canary) {
+      chain = chain.then(() => this.detectCanaryVersions());
+    } else {
+      // If no version bump was specified, check to ensure the working directory is clean
+      // and then attempt to publish all packages.
+      chain = chain
+          .then(() => this.verifyWorkingTreeClean())
+          .then(() => Array.from(this.packageGraph.values()))
+          .then(updates => {
+            const updatesVersions = updates.map(({ pkg }) => [pkg.name, pkg.version]);
+
+            return {
+              updates,
+              updatesVersions,
+              needsConfirmation: true,
+            };
+          });
+    }
+
+    return chain;
+  }
+}
+
+// New command meta data
+const travisPublishCmd = Object.assign({}, publishCmd, {
+  command: 'publish-travis [bump]',
+  describe: 'Publish all packages in the current project',
+  handler: factory
+});
+
+// Setup the custom cli
+function main(argv) {
+  // For lerna publication to work, the NPM token must be stored in the .npmrc file in the user home directory
+  if (process.env.TRAVIS && process.env.NPM_TOKEN) {
+    fs.writeFileSync(
+        path.resolve(process.env.HOME, '.npmrc'),
+        `//registry.npmjs.org/:_authToken=${process.env.NPM_TOKEN}\n`,
+        'utf8');
+  }
+
+  const context = {
+    lernaVersion: pkg.version,
+  };
+
+  return cli()
+      .command(addCmd)
+      .command(bootstrapCmd)
+      .command(changedCmd)
+      .command(cleanCmd)
+      .command(createCmd)
+      .command(diffCmd)
+      .command(execCmd)
+      .command(importCmd)
+      .command(initCmd)
+      .command(linkCmd)
+      .command(listCmd)
+      .command(publishCmd)
+      .command(runCmd)
+      .command(versionCmd)
+      .command(travisPublishCmd)
+      .parse(argv, context);
+}
+
+if (/publish/.test(process.argv[2])) {
+  // Looks like we're trying to publish packages
+  // Make sure the compiler version matches the package major version before publishing
+  const compilerVersionMatch = require(path.resolve(__dirname, 'version-match.js'));
+  const lernaConfig = require(path.resolve(__dirname, '..', 'lerna.json'));
+  const Compiler = require('google-closure-compiler').compiler;
+  const compiler = new Compiler({version: true});
+  compiler.run(function (exitCode, stdout) {
+    let versionInfo = (stdout || '').match(compilerVersionMatch);
+    versionInfo = versionInfo || [];
+    let packageVersion = new Semver(lernaConfig.version);
+    if (versionInfo.length < 2 || versionInfo[1] !== `v${packageVersion.major}`) {
+      console.log('Package major version does not match compiler version - skipping publication');
+      console.log(stdout);
+      process.exit(0);
+    } else {
+      // Invoke the cli with arguments
+      main(process.argv.slice(2));
+    }
+  });
+} else {
```

Comment:
I think this would be a bit clearer as

```javascript
if ('publish' == process.argv[2]) {
  // make sure we don't publish when the version in the JS compiler doesn't match what we expect.
 ...
}
main(process.argv.slice(2))
```

Reply:
I don't think that will work - the version check of the compiler is async so I need the else branch to prevent fall through.

Reply:
oops, I overlooked that.
Sorry.